runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java

```diff
@@ -18,76 +18,115 @@
 import java.io.IOException;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.Map;
 import org.apache.beam.runners.spark.aggregators.NamedAggregators;
 import org.apache.beam.runners.spark.util.BroadcastHelper;
+import org.apache.beam.runners.spark.util.SparkSideInputReader;
 import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.options.PipelineOptions;
 import org.apache.beam.sdk.transforms.Aggregator;
 import org.apache.beam.sdk.transforms.Combine;
 import org.apache.beam.sdk.transforms.OldDoFn;
 import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
 import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.transforms.windowing.WindowFn;
+import org.apache.beam.sdk.util.SideInputReader;
 import org.apache.beam.sdk.util.TimerInternals;
 import org.apache.beam.sdk.util.WindowedValue;
 import org.apache.beam.sdk.util.WindowingInternals;
+import org.apache.beam.sdk.util.WindowingStrategy;
 import org.apache.beam.sdk.util.state.InMemoryStateInternals;
 import org.apache.beam.sdk.util.state.StateInternals;
+import org.apache.beam.sdk.values.KV;
 import org.apache.beam.sdk.values.PCollectionView;
 import org.apache.beam.sdk.values.TupleTag;
 import org.apache.spark.Accumulator;
 import org.joda.time.Instant;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+
+
 /**
  * Spark runner process context.
  */
 public abstract class SparkProcessContext<InputT, OutputT, ValueT>
     extends OldDoFn<InputT, OutputT>.ProcessContext {
-
   private static final Logger LOG = LoggerFactory.getLogger(SparkProcessContext.class);
 
   private final OldDoFn<InputT, OutputT> fn;
   private final SparkRuntimeContext mRuntimeContext;
-  private final Map<TupleTag<?>, BroadcastHelper<?>> mSideInputs;
+  private final SideInputReader sideInputReader;
+  private final WindowingStrategy<?, ?> windowingStrategy;
 
-  protected WindowedValue<InputT> windowedValue;
+  WindowedValue<InputT> windowedValue;
 
   SparkProcessContext(OldDoFn<InputT, OutputT> fn,
-      SparkRuntimeContext runtime,
-      Map<TupleTag<?>, BroadcastHelper<?>> sideInputs) {
+                      SparkRuntimeContext runtime,
+                      Map<TupleTag<?>, KV<WindowingStrategy<?, ?>, BroadcastHelper<?>>> sideInputs,
+                      WindowingStrategy<?, ?> windowingStrategy) {
     fn.super();
     this.fn = fn;
     this.mRuntimeContext = runtime;
-    this.mSideInputs = sideInputs;
+    this.sideInputReader = new SparkSideInputReader(sideInputs);
+    this.windowingStrategy = windowingStrategy;
   }
 
   void setup() {
     setupDelegateAggregators();
   }
 
+  Iterable<ValueT> callWithCtxt(Iterator<WindowedValue<InputT>> iter) throws Exception{
+    this.setup();
+    // skip if bundle is empty.
+    if (!iter.hasNext()) {
+      return Lists.newArrayList();
+    }
+    try {
+      fn.setup();
+      fn.startBundle(this);
+      return this.getOutputIterable(iter, fn);
```

Comment:
Does this factoring mean that there might be a single `startBundle` call followed by multiple iterations, thus multiple `finishBundle` calls?